src/react-components/input/Slider.js

```diff
@@ -0,0 +1,143 @@
+    ")"
+  ].join("");
+}
+
+function updateGradient(el, min, max, value) {
+  if (el) {
+    const ratio = ((value - min) / (max - min)).toFixed(2);
+    el.style.backgroundImage = getLinearGradientCSS(ratio, trackColor, "transparent");
+  }
+}
+
+export function Slider({ min, max, step, digits, value, defaultValue, onChange, disabled, className }) {
+  const myRoot = useRef();
+  const inputRef = useRef();
+  const [displayValue, setDisplayValue] = useState(-1);
+  const [isDragging, setIsDragging] = useState(false);
+
+  let currentValue = defaultValue;
+  if (isDragging) {
+    currentValue = displayValue !== -1 ? displayValue : defaultValue;
+  } else {
+    currentValue = value !== undefined ? value : displayValue !== -1 ? displayValue : defaultValue;
+  }
+
+  const stopDragging = useCallback(
+    () => {
+      setIsDragging(false);
+    },
+    [setIsDragging]
+  );
+
+  const onUpdateValue = useCallback(
+    v => {
+      setDisplayValue(v);
+      if (onChange) {
+        onChange(v);
+      }
+    },
+    [onChange, setDisplayValue]
+  );
+
+  const drag = useCallback(
+    e => {
+      if (!isDragging) return;
+      const viewportOffset = myRoot.current.getBoundingClientRect();
+      const t = Math.max(0, Math.min((e.clientX - viewportOffset.left) / viewportOffset.width, 1));
+      const num = round(step, min + t * (max - min));
+      onUpdateValue(num);
+    },
+    [isDragging, step, min, max, onUpdateValue]
+  );
+
+  useEffect(
+    () => {
+      window.addEventListener("mouseup", stopDragging);
+      window.addEventListener("mousemove", drag);
+
+      updateGradient(inputRef.current, min, max, currentValue);
+
+      return () => {
+        window.removeEventListener("mouseup", stopDragging);
+        window.removeEventListener("mousemove", drag);
+      };
+    },
+    [drag, stopDragging, inputRef, min, max, currentValue]
+  );
+
+  return (
+    <div className={classNames(styles.numberWithRange, className)}>
+      <div
+        ref={myRoot}
+        className={classNames(styles.rangeSlider)}
+        disabled={disabled}
+        onMouseDown={e => {
+          e.preventDefault();
+          if (disabled) return;
+          setIsDragging(true);
+          const viewportOffset = myRoot.current.getBoundingClientRect();
+          const t = Math.max(0, Math.min((e.clientX - viewportOffset.left) / viewportOffset.width, 1));
+          const num = round(step, min + t * (max - min));
+          const finalNum = parseFloat(num.toFixed(digits));
+          onUpdateValue(finalNum);
+        }}
+      >
+        <input
+          ref={inputRef}
+          type="range"
+          step={step}
+          min={min}
+          max={max}
+          value={currentValue}
+          disabled={disabled}
+          onChange={e => {
+            if (disabled) return;
+            const num = round(step, parseFloat(e.target.value));
+            const finalNum = parseFloat(num.toFixed(digits));
```

Comment:
This widget doesn't actually display a number so it feels weird for it to b doing this truncation. Even if it did display a number it seems like that should only be done in the display logic, not in the actual value it outputs. We also seem to do this again outside of the component when using it.

Reply:
same